File: polyglot/bridges/julia/bridge.jl
#!/usr/bin/env julia
# WhiteMagic Julia JSON stdio bridge
# Reads JSON requests from stdin, writes JSON responses to stdout

using JSON

function handle(req)
    method = req["method"]
    p = req["params"]

    if method == "ping"
        return Dict("status" => "ok", "backend" => "julia")

    elseif method == "encode"
        coord = HolographicMemory.encode_5d(p["text"])
        return Dict(
            "status" => "ok",
            "x" => coord[1], "y" => coord[2], "z" => coord[3],
            "w" => coord[4], "v" => coord[5],
            "zone" => HolographicMemory.zone_of(coord)
        )

    elseif method == "nearest_neighbors"
        query = HolographicMemory.encode_5d(p["query"])
        coords = [HolographicMemory.encode_5d(t) for t in p["texts"]]
        k = get(p, "k", 3)
        # Use KD-tree for large coordinate sets (O(log n) vs O(n))
        if length(coords) >= 64
            result = HolographicMemory.nearest_neighbors_kdtree(query, coords, k)
        else
            result = HolographicMemory.nearest_neighbors(query, coords, k)
        end
        return Dict(
            "status" => "ok",
            "results" => [Dict("index" => r[1], "distance" => r[2]) for r in result]
        )

    elseif method == "constellation_detect"
        coords = [HolographicMemory.encode_5d(t) for t in p["texts"]]
        threshold = get(p, "threshold", 0.8)
        min_size = get(p, "min_size", 2)
        clusters = HolographicMemory.constellation_detect(coords, threshold, min_size)
        return Dict(
            "status" => "ok",
            "clusters" => [c for c in clusters]
        )

    elseif method == "coherence"
        coords = [HolographicMemory.encode_5d(t) for t in p["texts"]]
        score = HolographicMemory.coherence_score(coords)
        return Dict("status" => "ok", "coherence" => score)

    else
        return Dict("status" => "error", "error" => "Unknown method: $method")
    end
end

# Main loop: read JSON line, handle, write JSON line

# Warmup: trigger JIT for hot paths (reduces first-request latency from 9.5s to 3s)
HolographicMemory.encode_5d("warmup")
c1 = HolographicMemory.encode_5d("w")
c2 = HolographicMemory.encode_5d("x")
HolographicMemory.nearest_neighbors(c1, [c2], 1)
HolographicMemory.zone_of(c1)
# Warmup KD-tree path with enough points to trigger it
many_coords = [HolographicMemory.encode_5d("warmup_$i") for i in 1:70]
HolographicMemory.nearest_neighbors_kdtree(c1, many_coords, 5)

while !eof(stdin)
    line = readline(stdin)
    isempty(line) && continue
    try
        req = JSON.parse(line)
        resp = handle(req)
        println(stdout, JSON.json(resp))
        flush(stdout)
    catch e
        err = Dict("status" => "error", "error" => string(e))
        println(stdout, JSON.json(err))
        flush(stdout)
    end
end

File: polyglot/bridges/elixir/bridge.exs
# WhiteMagic Elixir JSON stdio bridge
# Uses only Coordinate pure functions (no GenServer needed)

defmodule WhiteMagic.Bridge do
  alias WhiteMagicElixir.Coordinate

  def handle(req) do
    method = req["method"]
    p = req["params"]

    case method do
      "ping" ->
        %{status: "ok", backend: "elixir"}

      "encode" ->
        coord = Coordinate.encode(p["text"])
        %{
          status: "ok",
          x: coord.x, y: coord.y, z: coord.z,
          w: coord.w, v: coord.v,
          zone: Coordinate.zone(coord)
        }

      "nearest_neighbors" ->
        query = Coordinate.encode(p["query"])
        coords = Enum.map(p["texts"], &Coordinate.encode/1)
        k = p["k"] || 3
        result = nearest_neighbors(query, coords, k)
        %{
          status: "ok",
          results: Enum.map(result, fn {idx, dist} ->
            %{index: idx, distance: dist}
          end)
        }

      "distance" ->
        a = Coordinate.encode(p["a"])
        b = Coordinate.encode(p["b"])
        %{status: "ok", distance: Coordinate.distance(a, b)}

      "coherence" ->
        coords = Enum.map(p["texts"], &Coordinate.encode/1)
        %{status: "ok", coherence: Coordinate.coherence(coords)}

      _ ->
        %{status: "error", error: "Unknown method: #{method}"}
    end
  end

  defp nearest_neighbors(query, coords, k) do
    coords
    |> Enum.with_index()
    |> Enum.map(fn {c, i} -> {i, Coordinate.distance(query, c)} end)
    |> Enum.sort_by(fn {_, d} -> d end)
    |> Enum.take(k)
  end

  def loop do
    case IO.read(:line) do
      :eof -> :ok
      line when is_binary(line) ->
        text = String.trim(line)
        if text == "" do
          loop()
        else
          try do
            req = Jason.decode!(text)
            resp = handle(req)
            IO.puts(Jason.encode!(resp))
          rescue
            e ->
              err = %{status: "error", error: Exception.message(e)}
              IO.puts(Jason.encode!(err))
          end
          loop()
        end
    end
  end
end

# Warmup: trigger module loading before serving requests
WhiteMagicElixir.Coordinate.encode("warmup")

WhiteMagic.Bridge.loop()
